Support stop msg for daemon node

gmarzot opened the pull request:
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Proposed changes

This PR adds support to the node-red-node-daemon for a stop message. When stopped the process will be terminated as per the configured signal and will not be restarted by redo until re-set via a subsequent start message. 

Documentation updated on start message to indicate it will not restart a running process


- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red-nodes/blob/master/CONTRIBUTING.md)
- [x] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [x] I have run `grunt` to verify the unit tests pass
- [x] I have added suitable unit tests to cover the new/changed functionality

dceejay commented on utility/daemon/daemon.js:
```diff
@@ -32,16 +33,30 @@ module.exports = function(RED) {
+                            args = parseArgs(msg.args.trim());
+                        }
+                        runit(args);
+                    } else {
```
Why are we doing this here ? If it is called with start and is already running then we don't want to kill it do we ? Surely just leave it running.

dceejay commented on utility/daemon/daemon.js:
```diff
@@ -32,16 +33,30 @@ module.exports = function(RED) {
+                    if (node.running) {
+                        node.child.kill(node.closer);
+                    }
+                    node.debug(node.cmd+" stopped by msg");
```
Don't think we need extra debug.

gmarzot commented:
i added the kill on the start msg to trigger a restart, to match documented
behavior..

but equally valid (or more so i guess) would be to change the doc to say
that msg.start only starts (no restart behavior). I can make that change.

eh... debug to cover each message type.. i see no harm as it is off
normally.. your call.

-G

On Thu, Jul 20, 2023 at 5:13 PM Dave Conway-Jones ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In utility/daemon/daemon.js
> <https://github.com/node-red/node-red-nodes/pull/1018#discussion_r1269972333>
> :
>
> >                      if (typeof msg.kill !== "string" || msg.kill.length === 0 || !msg.kill.toUpperCase().startsWith("SIG") ) { msg.kill = "SIGINT"; }
>                      node.child.kill(msg.kill.toUpperCase());
>                  }
> -                else if (msg.hasOwnProperty("start") && !node.running) {
> -                    let args = "";
> -                    if (msg.hasOwnProperty("args") && msg.args.length > 0) {
> -                        args = parseArgs(msg.args.trim());
> +                else if (msg.hasOwnProperty("start")) {
> +                    node.stopped = false;
> +                    if (!node.running) {
> +                        let args = "";
> +                        if (msg.hasOwnProperty("args") && msg.args.length > 0) {
> +                            args = parseArgs(msg.args.trim());
> +                        }
> +                        runit(args);
> +                    } else {
>
> Why are we doing this here ? If it is called with start and is already
> running then we don't want to kill it do we ? Surely just leave it running.
> ------------------------------
>
> In utility/daemon/daemon.js
> <https://github.com/node-red/node-red-nodes/pull/1018#discussion_r1269972571>
> :
>
> > @@ -32,16 +33,30 @@ module.exports = function(RED) {
>
>          function inputlistener(msg) {
>              if (msg != null) {
> -                if (msg.hasOwnProperty("kill") && node.running) {
> +                if (msg.hasOwnProperty("stop")) {
> +                    node.stopped = true;
> +                    if (node.running) {
> +                        node.child.kill(node.closer);
> +                    }
> +                    node.debug(node.cmd+" stopped by msg");
>
> Don't think we need extra debug.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/node-red/node-red-nodes/pull/1018#pullrequestreview-1540102002>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB7T3ICTCHIIJYMMHGAIGBLXRGNOZANCNFSM6AAAAAA2RV4ODM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

gmarzot commented:
one lingering thought is that if someone wants to add new args and trigger
a  restart.. the current PR supports that w/ a single msg...

but that can be accomplished with stop then start w/ args.. so i am
continuing to remove the kill i added to msg.start

On Thu, Jul 20, 2023 at 5:27 PM Giovanni Marzot ***@***.***> wrote:

> i added the kill on the start msg to trigger a restart, to match
> documented behavior..
>
> but equally valid (or more so i guess) would be to change the doc to say
> that msg.start only starts (no restart behavior). I can make that change.
>
> eh... debug to cover each message type.. i see no harm as it is off
> normally.. your call.
>
> -G
>
> On Thu, Jul 20, 2023 at 5:13 PM Dave Conway-Jones <
> ***@***.***> wrote:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In utility/daemon/daemon.js
>> <https://github.com/node-red/node-red-nodes/pull/1018#discussion_r1269972333>
>> :
>>
>> >                      if (typeof msg.kill !== "string" || msg.kill.length === 0 || !msg.kill.toUpperCase().startsWith("SIG") ) { msg.kill = "SIGINT"; }
>>                      node.child.kill(msg.kill.toUpperCase());
>>                  }
>> -                else if (msg.hasOwnProperty("start") && !node.running) {
>> -                    let args = "";
>> -                    if (msg.hasOwnProperty("args") && msg.args.length > 0) {
>> -                        args = parseArgs(msg.args.trim());
>> +                else if (msg.hasOwnProperty("start")) {
>> +                    node.stopped = false;
>> +                    if (!node.running) {
>> +                        let args = "";
>> +                        if (msg.hasOwnProperty("args") && msg.args.length > 0) {
>> +                            args = parseArgs(msg.args.trim());
>> +                        }
>> +                        runit(args);
>> +                    } else {
>>
>> Why are we doing this here ? If it is called with start and is already
>> running then we don't want to kill it do we ? Surely just leave it running.
>> ------------------------------
>>
>> In utility/daemon/daemon.js
>> <https://github.com/node-red/node-red-nodes/pull/1018#discussion_r1269972571>
>> :
>>
>> > @@ -32,16 +33,30 @@ module.exports = function(RED) {
>>
>>          function inputlistener(msg) {
>>              if (msg != null) {
>> -                if (msg.hasOwnProperty("kill") && node.running) {
>> +                if (msg.hasOwnProperty("stop")) {
>> +                    node.stopped = true;
>> +                    if (node.running) {
>> +                        node.child.kill(node.closer);
>> +                    }
>> +                    node.debug(node.cmd+" stopped by msg");
>>
>> Don't think we need extra debug.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/node-red/node-red-nodes/pull/1018#pullrequestreview-1540102002>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AB7T3ICTCHIIJYMMHGAIGBLXRGNOZANCNFSM6AAAAAA2RV4ODM>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

dceejay commented:
Thanks for the PR.
